src/System.Memory/src/System/Buffers/ArrayMemoryPool.ArrayMemoryPoolBuffer.cs

```diff
@@ -0,0 +1,92 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.InteropServices;
+using System.Threading;
+
+namespace System.Buffers
+{
+    internal sealed partial class ArrayMemoryPool<T> : MemoryPool<T>
+    {
+        private sealed class ArrayMemoryPoolBuffer : OwnedMemory<T>
+        {
+            private T[] _array;
+            private bool _disposed;
+            private int _referenceCount;
+
+            public ArrayMemoryPoolBuffer(int size)
+            {
+                _array = ArrayPool<T>.Shared.Rent(size);
+            }
+
+            public sealed override int Length => _array.Length;
+
+            public sealed override bool IsDisposed => _disposed;
+
+            protected sealed override bool IsRetained => _referenceCount > 0;
+
+            public sealed override Span<T> Span
+            {
+                get
+                {
+                    if (IsDisposed)
+                        ThrowHelper.ThrowObjectDisposedException(nameof(ArrayMemoryPoolBuffer));
+
+                    return _array;
+                }
+            }
+
+            protected sealed override void Dispose(bool disposing)
+            {
+                T[] array = Interlocked.Exchange(ref _array, null);
```

Comment:
What if it's currently retained?

Reply:
Then it throws an InvalidOperationException. That check is encapsulated in the OwnedMemory class itself.

Reply:
What throws an InvalidOperationException?

Reply:
Right here: https://github.com/dotnet/corefx/blob/83f35ef9a79b5cd1412aeaad8d2cb84e1bfe73c4/src/Common/src/CoreLib/System/Buffers/OwnedMemory.cs#L63

Reply:
But you're not calling base.Dispose(disposing).  So who calls that?

Reply:
Oh, sorry, different overload.  I see.